gpkit/constraint_meta.py

```diff
@@ -0,0 +1,123 @@
+from abc import ABCMeta, abstractmethod, abstractproperty
+
+
+class Constraint(object):
+    __metaclass__ = ABCMeta
+
+    @abstractproperty
+    def varkeys(self):
+        "Varkeys present in the Constraint"
+        pass
+
+    @abstractproperty
+    def substitutions(self):
+        "Substitutions that the Constraint currently knows about"
+        pass
+
+    @substitutions.setter
+    def substitutions(self, newvalue):
+        ".substitutions mist be settable"
+        pass
+
+    @abstractmethod
+    def __str__(self):
+        "string representation of the constraint, used for printing programs"
+        pass
+
+    @abstractmethod
+    def latex(self):
+        "latex representation of the constraint, used for printing programs"
+        pass
+
+    def _repr_latex_(self):
+        "Classes that inherit from Constraint will show as latex in IPython"
+        return "$$" + self.latex() + "$$"
+
+    @abstractmethod
+    def sub(self, subs, value=None):
+        "Returns new object without any of the varkeys in subs.keys()"
```

Comment:
cool. and what happens with `self.substitutions` in cases where there is/is not an overlap with `subs`?